lib/client/api.go

```diff
@@ -462,7 +462,19 @@ func RetryWithRelogin(ctx context.Context, tc *TeleportClient, fn func() error)
 		return trace.Wrap(err)
 	}
 	if err := tc.ActivateKey(ctx, key); err != nil {
-		return trace.Wrap(err)
+		if len(tc.JumpHosts) == 0 {
+			return trace.Wrap(err)
+		}
+		errViaJumphost := err
+		// ActivateKey re-fetches the list of CAs from auth server. If
+		// JumpHosts was pointing at the leaf cluster, this could've caused the
+		// above error. Try to ActivateKey without JumpHosts to force it to use
+		// the root cluster.
+		if err := tc.WithoutJumpHosts(func(tc *TeleportClient) error {
+			return tc.ActivateKey(ctx, key)
+		}); err != nil {
+			return trace.NewAggregate(errViaJumphost, err)
+		}
```

Comment:
Even though this should work, I don't like this general approach. `WithoutJumpHosts` is wonky, we should focus on when jumphosts *are* supposed to be used, i.e. take a `WithJumpHosts` approach.

Also if we're going down this way, this should probably be put directly into `TeleportClient.UpdateTrustedCA`, not ad hoc in `RetryWithRelogin`.

Reply:
yeah, it's pretty ugly.
the whole client library is a spaghetti and I don't know where jumphosts really matter and where not.
moved the retry into `ActivateKey` for now.